File: src/main/kotlin/xyz/mahmoudahmed/dsl/NestedSourcePathBuilder.java
package xyz.mahmoudahmed.dsl;


import java.util.function.Function;

public class NestedSourcePathBuilder {
    private final NestedMappingBuilder parent;
    private final String sourcePath;
    private Function<Object, Object> transformation;

    NestedSourcePathBuilder(NestedMappingBuilder parent, String sourcePath) {
        this.parent = parent;
        this.sourcePath = sourcePath;
    }

    public NestedTargetPathBuilder to(String targetPath) {
        return new NestedTargetPathBuilder(parent, sourcePath, targetPath, transformation);
    }

    public NestedSourcePathBuilder transform(Function<Object, Object> transformation) {
        this.transformation = transformation;
        return this;
    }
}

File: src/main/kotlin/xyz/mahmoudahmed/dsl/collections/NestedSourcePathBuilder.kt
package xyz.mahmoudahmed.dsl.collections

/**
 * Builder for configuring a source path in a nested mapping.
 */
class NestedSourcePathBuilder(
    private val parent: NestedMappingBuilder,
    private val sourcePath: String
) {
    private var transformation: ((Any) -> Any)? = null

    /**
     * Specifies the target path for the mapping.
     *
     * @param targetPath Path where the value will be set
     * @return Builder for finalizing the mapping
     */
    fun to(targetPath: String): NestedTargetPathBuilder {
        return NestedTargetPathBuilder(parent, sourcePath, targetPath, transformation)
    }

    /**
     * Applies a transformation to the source value.
     *
     * @param transformation Function to transform the source value
     * @return This builder for chaining
     */
    fun transform(transformation: Any): NestedSourcePathBuilder {
        @Suppress("UNCHECKED_CAST")
        this.transformation = transformation as (Any) -> Any
        return this
    }
}
